CARDS-2221: Create a script within the CARDS repository for building a self-contained Docker image using the cards-sling-feature-downloader utility

IntegralProgrammer opened the pull request:
This Pull Request adds the `Utilities/Packaging/Docker/build_self_contained.sh` script (and its dependencies) for easily building a self-contained CARDS Docker image. Many of the scripts included in this Pull Request are slightly modified versions of the ones in the https://github.com/data-team-uhn/cards-sling-feature-downloader (private) GitHub repository.

Testing Instructions
--------------------------

1. Ensure that you have built a `cards/sling-feature-downloader` Docker image from the https://github.com/data-team-uhn/cards-sling-feature-downloader (private) GitHub repository.
2. `cd Utilities/Packaging/Docker`
3. `./build_self_contained.sh cards/cards-self-contained`. Ensure that the `cards/cards-self-contained` Docker image is built without error.
4. `cd ../../../compose-cluster`
5. `python3 generate_compose_yaml.py --oak_filesystem --cards_project cards4proms --composum`
6. Edit `docker-compose.yml` and change `image: cards/cards:latest` to `image: cards/cards-self-contained:latest`
7. `docker-compose build && docker-compose up -d`
8. Ensure that CARDS starts and is available on http://localhost:8080
9. Clean up with `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`

IntegralProgrammer committed:
CARDS-2221: Create a script within the CARDS repository for building a self-contained Docker image using the cards-sling-feature-downloader utility

IntegralProgrammer committed:
CARDS-2221: Create a script within the CARDS repository for building a self-contained Docker image using the cards-sling-feature-downloader utility

First check that the cards/sling-feature-downloader Docker image exists on the local machine before proceeding with the rest of the build_self_contained.sh script

IntegralProgrammer committed:
CARDS-2221: Create a script within the CARDS repository for building a self-contained Docker image using the cards-sling-feature-downloader utility

"mvn clean install -Pdocker" will create a Docker image named "cards/cards:latest" if this repository is checked out to a version of CARDS that ends in "-SNAPSHOT". Otherwise, the version of CARDS that this repository is checked out to (eg. "0.9.7", "0.9.8", etc...) will be used for naming the CARDS Docker image (eg. "cards/cards:0.9.7", "cards/cards:0.9.8", etc...)

IntegralProgrammer committed:
CARDS-2221: Create a script within the CARDS repository for building a self-contained Docker image using the cards-sling-feature-downloader utility

Display an error if the cards/sling-feature-downloader Docker image is not present on the local machine

sdumitriu commented on Utilities/Packaging/Docker/build_self_contained.sh:
```diff
@@ -0,0 +1,56 @@
+OUTPUT_DOCKER_IMAGE=$1
+
+# First check that the cards/sling-feature-downloader Docker image exists on the local machine
+(docker image inspect cards/sling-feature-downloader > /dev/null) || { echo "Fail: The cards/sling-feature-downloader Docker image does not exist. Exiting."; exit -1; }
```
Add instructions for getting it.

IntegralProgrammer replied:
Added instructions for obtaining the `cards/sling-feature-downloader` Docker image in https://github.com/data-team-uhn/cards/pull/1488/commits/dab9818f76c2995ac3073b80f142a2fe16c56354 - the [data-team-uhn/cards-sling-feature-downloader](https://github.com/data-team-uhn/cards-sling-feature-downloader) repository is now _public_ and uses [a custom release of org.apache.sling.feature.launcher - `1.3.0.cards.1`](https://github.com/data-team-uhn/sling-org-apache-sling-feature-launcher/releases/tag/org.apache.sling.feature.launcher-1.3.0.cards.1) to gather all required JARs.

If this script is executed on a non-release (ie. the version ends with `-SNAPSHOT`), Webpack will build in development mode, otherwise Webpack will run in production mode - https://github.com/data-team-uhn/cards/pull/1488/commits/f94681fdcb27443ca62da0d258979e3ad25bd49c.

A safety check has also been added to the `build_self_contained.sh` script - https://github.com/data-team-uhn/cards/pull/1488/commits/868995becba3839f58b4f5f613a5f89453e686bd.

IntegralProgrammer committed:
CARDS-2221: Create a script within the CARDS repository for building a self-contained Docker image using the cards-sling-feature-downloader utility

If this is a release that we are building (ie. a version that does not end with "-SNAPSHOT") build using the production Maven arguments

IntegralProgrammer committed:
CARDS-2221: Create a script within the CARDS repository for building a self-contained Docker image using the cards-sling-feature-downloader utility

Check that the user has supplied an argument for OUTPUT_DOCKER_IMAGE

IntegralProgrammer committed:
CARDS-2221: Create a script within the CARDS repository for building a self-contained Docker image using the cards-sling-feature-downloader utility

Add instructions for obtaining the cards/sling-feature-downloader Docker image
